### PR TITLE
Kinder replace minikube/namnx228

### PR DESCRIPTION
**Pull request by namnx228**
In this pull request, I add `kinder ` as another option to create K8S. `kinder` is necessary for CI as it is more lightweight and runs faser than  `minikube` . 

By default, `minikube` is still used to deploy K8S in the dev-env.

Note: This PR is developed on top of PR: `improvement/namnx228-download-1-place`. Therefore, please review that branch before this one.

### CODE REVIEW

**Comment by russellb** [comment]
Thanks for the PR!  I don't mean to ignore this, I'm just waiting until we finish #37 since this depends on it

**Comment by nordixinfra** [comment]
Can one of the admins verify this patch?

**Comment by russellb** [comment]
It looks like this needs a rebase, and then i'll try this out, thanks!

**Comment by maelk** [comment]
@russellb this is now rebased, thanks

**Comment by russellb** [comment]
sorry, i introduced a conflict in the Makefile.  I tried to fix it but can't push to the branch.  (github gives you the choice to allow maintainers to push to your branch for a PR)

File: 01_install_requirements.sh [hunk 1]
```diff
@@ -59,3 +49,51 @@ ANSIBLE_FORCE_COLOR=true ansible-playbook \
...
+      chmod +x docker-machine-driver-kvm2
+      sudo mv docker-machine-driver-kvm2 /usr/local/bin/.
+  fi
+else
+# Install kinder to replace minikube
+  if [ ! -d ${GOPATH}/src/kubeadm ]; then
+    pushd ${GOPATH}/src/
+    git clone https://github.com/kubernetes/kubeadm.git
+    pushd kubeadm
+    popd
```
Comment (russellb) [inline 01_install_requirements.sh]: looks like these last 2 lines don't do anything

File: 01_install_requirements.sh [hunk 2]
```diff
@@ -8,18 +9,7 @@ else
...
-    sudo mv minikube /usr/local/bin/.
-fi
-
-if ! which docker-machine-driver-kvm2 2>/dev/null ; then
-    curl -LO https://storage.googleapis.com/minikube/releases/latest/docker-machine-driver-kvm2
-    chmod +x docker-machine-driver-kvm2
-    sudo mv docker-machine-driver-kvm2 /usr/local/bin/.
-fi
-
+
```
Comment (russellb) [inline 01_install_requirements.sh]: extra blank line

File: 01_install_requirements.sh [hunk 3]
```diff
@@ -59,3 +49,51 @@ ANSIBLE_FORCE_COLOR=true ansible-playbook \
...
+    pushd ${GOPATH}/src/
+    git clone https://github.com/kubernetes/kubeadm.git
+    pushd kubeadm
+    popd
+    popd
+  fi
+  
+  # if [ ! -d ${GOPATH}/src/sigs.k8s.io/kind ]; then
+  #   GO111MODULE="on" go get -u sigs.k8s.io/kind@v0.4.0
+  # fi
```
Comment (russellb) [inline 01_install_requirements.sh]: want to remove this?  or add a comment about why it's here?

File: 01_install_requirements.sh [hunk 4]
```diff
@@ -59,3 +49,51 @@ ANSIBLE_FORCE_COLOR=true ansible-playbook \
...
+  
+  if [ ! -f ${GOPATH}/bin/kinder ]; then
+    pushd $HOME/go/src/kubeadm/kinder
+    GO111MODULE=on go install
+    popd
+  fi
+  
+  sudo docker pull kindest/node:v1.15.0
+  
+  # sudo usermod -aG docker $USER
```
Comment (russellb) [inline 01_install_requirements.sh]: want to remove this?  or add a comment about why it's here?

File: 02_configure_host.sh [hunk 5]
```diff
@@ -63,14 +63,14 @@ else
       if [ ! -e /etc/sysconfig/network-scripts/ifcfg-provisioning ] ; then
           echo -e "DEVICE=provisioning\nTYPE=Bridge\nONBOOT=yes\nNM_CONTROLLED=no\nBOOTPROTO=static\nIPADDR=172.22.0.1\nNETMASK=255.255.255.0" | sudo dd of=/etc/sysconfig/network-scripts/ifcfg-provisioning
       fi
-      sudo ifdown provisioning || true
-      sudo ifup provisioning
+      sudo ip link set provisioning down || true
+      sudo ip link set provisioning up || true
```
Comment (russellb) [inline 02_configure_host.sh]: Why is this changed?

Comment (russellb) [reply 02_configure_host.sh]: same question about the other places where `ifdown`/`ifup` were changed to `ip link set up` / `down`

File: 03_launch_mgmt_cluster.sh [hunk 6]
```diff
@@ -79,6 +79,20 @@ function launch_minikube() {
     minikube start
 }
 
+function launch_kinder() {
+  kinder="${GOPATH}/bin/kinder"
```
Comment (russellb) [inline 03_launch_mgmt_cluster.sh]: shouldn't this already be in `$PATH` ?

File: 03_launch_mgmt_cluster.sh [hunk 7]
```diff
@@ -79,6 +79,20 @@ function launch_minikube() {
...
+  if [[ $(sudo $kinder get clusters) == "" ]]; then
+    sudo $kinder create cluster
+    sudo $kinder do kubeadm-init
+    kinder_kubeconfig_path="$(sudo $kinder get kubeconfig-path)"
+    sudo rm -f ${HOME}/.kube/config
+    mkdir -p ${HOME}/.kube/
+    sudo mv $kinder_kubeconfig_path ${HOME}/.kube/config
+    sudo chown ${USER}:${USER} ${HOME}/.kube/config
+    #TODO: Attach `provisioning` interface to node which runs Ironic service.
```
Comment (russellb) [inline 03_launch_mgmt_cluster.sh]: not sure I understand the TODO here -- Ironic is running on the host where the `provisioning` bridge is located -- can you expand on what's left to do here?

**Comment by namnx228** [comment]
Hi Russelb, I have enabled that permission for you.

File: config_example.sh [hunk 8]
```diff
@@ -57,6 +57,9 @@
 #
 #export CAPBM_RUN_LOCAL=true
 
+#Set Kubenetes cluster
```
Comment (russellb) [inline config_example.sh]: Can you add a comment about what the choices are?

default: `minikube`, also supported: `kinder` ?

File: lib/common.sh [hunk 9]
```diff
@@ -2,6 +2,7 @@
 
 eval "$(go env)"
 
+
```
Comment (russellb) [inline lib/common.sh]: extra blank line added here

File: vm-setup/roles/libvirt/defaults/main.yml [hunk 10]
```diff
@@ -23,7 +23,5 @@ extradisks_list:
 # size of the disks to create when using extradisks
 extradisks_size: 8G
 
-
-
```
Comment (russellb) [inline vm-setup/roles/libvirt/defaults/main.yml]: unrelated changes

**Comment by maelk** [comment]
Please ignore last ci failure. it is unrelated to the patchset

**Comment by namnx228** [comment]
@russellb , thanks for your review, I will update the fix soon

File: 01_install_requirements.sh (hunk 1 above)
Comment (namnx228) [reply 01_install_requirements.sh]: Ok, I will remove them

File: 01_install_requirements.sh (hunk 2 above)
Comment (namnx228) [reply 01_install_requirements.sh]: Solved!

File: 01_install_requirements.sh (hunk 3 above)
Comment (namnx228) [reply 01_install_requirements.sh]: removed it!

File: 01_install_requirements.sh (hunk 4 above)
Comment (namnx228) [reply 01_install_requirements.sh]: removed!!

File: 03_launch_mgmt_cluster.sh (hunk 6 above)
Comment (namnx228) [reply 03_launch_mgmt_cluster.sh]: Yes, but not in $PATH of root user. Therefore, I use the full path to run Kinder with sudo right.

File: 03_launch_mgmt_cluster.sh (hunk 7 above)
Comment (namnx228) [reply 03_launch_mgmt_cluster.sh]: Sorry, my mistake. 
Previously, I thought that if we move Ironic to run inside a container, we will need `provisioning` bridge to make it visible to the network outside the cluster. But it seems to be my misunderstanding.

File: config_example.sh (hunk 8 above)
Comment (namnx228) [reply config_example.sh]: Done

File: 02_configure_host.sh (hunk 5 above)
Comment (namnx228) [reply 02_configure_host.sh]: ifdown/ifup doesn't work without net-tools package in Ubuntu, so and just change use ip. It work fine on all linux distributions, and we don't need to install additional packages for it.

Comment (russellb) [reply 02_configure_host.sh]: I don't think those commands are equivalent, though?

Comment (namnx228) [reply 02_configure_host.sh]: Yeah, but in context of dev-env, I don't see any problems when we use `ip link set`. 
Do you think are there any problems with it ? But If you really think that this change is not problematic, I will drop it in this pull request.

**Comment by namnx228** [comment]
This pull request is still in testing process

File: 01_prepare_host.sh [hunk 11]
```diff
@@ -13,19 +13,7 @@ else
   # shellcheck disable=SC1091
   source centos_install_requirements.sh
 fi
-
-if ! which minikube 2>/dev/null ; then
```
Comment (maelk) [inline 01_prepare_host.sh]: The access rights on this file have changed!

Comment (maelk) [reply 01_prepare_host.sh]: That's why the CI failed

Comment (namnx228) [reply 01_prepare_host.sh]: What is the original access rights on this file ?

Comment (namnx228) [reply 01_prepare_host.sh]: Fixed. Please trigger the test

File: 02_configure_host.sh (hunk 5 above)
Comment (russellb) [reply 02_configure_host.sh]: I'd at least rather see it in a separate PR since it doesn't seem related to adding kinder support

File: 03_launch_mgmt_cluster.sh (hunk 6 above)
Comment (russellb) [reply 03_launch_mgmt_cluster.sh]: ah, OK, makes sense.

File: config_example.sh (hunk 8 above)
Comment (russellb) [reply config_example.sh]: looks like an addition is still needed to clarify which one is the default
